### PR TITLE
Map over notebook task

### PR DESCRIPTION
**Pull request by pingsutw**
# TL;DR
This PR allows map over notebook task

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
```python
import os
import pathlib
import typing

from flytekit import kwtypes, task, workflow, ImageSpec, map_task
from flytekitplugins.papermill import NotebookTask

new_flytekit = "git+https://github.com/flyteorg/flytekit@022ade936d2fe1eb6a2d705c319cce7d2c66ade6"
new_papermill = "git+https://github.com/flyteorg/flytekit.git@022ade936d2fe1eb6a2d705c319cce7d2c66ade6#subdirectory=plugins/flytekit-papermill"
image = ImageSpec(registry="pingsutw", packages=[new_papermill, new_flytekit], apt_packages=["git"])

nb = NotebookTask(
    name="simple-nb",
    notebook_path=os.path.join(
        pathlib.Path(__file__).parent.absolute(), "nb-simple.ipynb"
    ),
    render_deck=True,
    inputs=kwtypes(a=float),
    outputs=kwtypes(square=float),
    container_image=image,
)


@workflow
def wf(a: float) -> typing.List[float]:
    return map_task(nb)(a=[a])


if __name__ == "__main__":
    print(wf(a=3.14))
```

image

## Tracking Issue
https://flyte-org.slack.com/archives/CP2HDHKE1/p1684471512753359

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit/core/map_task.py [hunk 1]
```diff
@@ -64,7 +64,7 @@ def __init__(
         else:
             actual_task = python_function_task
 
-        if not isinstance(actual_task, PythonFunctionTask):
+        if not isinstance(actual_task, PythonTask) or not issubclass(type(actual_task), PythonInstanceTask):
```
Comment (kumare3) [inline flytekit/core/map_task.py:67]: You don't need the second check instance is subclass of pythontask

File: flytekit/core/map_task.py [hunk 2]
```diff
@@ -76,7 +76,11 @@ def __init__(
 
         collection_interface = transform_interface_to_list_interface(actual_task.python_interface, self._bound_inputs)
         self._run_task: PythonFunctionTask = actual_task
-        _, mod, f, _ = tracker.extract_task_module(actual_task.task_function)
+        if hasattr(actual_task, "_IMPLICIT_OP_NOTEBOOK_TYPE"):
```
Comment (kumare3) [inline flytekit/core/map_task.py:79]: This seems to special for notebook- we should make it for instance tasks in general

File: plugins/flytekit-papermill/flytekitplugins/papermill/task.py [hunk 3]
```diff
@@ -277,16 +270,14 @@ def execute(self, **kwargs) -> Any:
         output_list = []
 
         for k, type_v in self.python_interface.outputs.items():
-            if k == self._IMPLICIT_OP_NOTEBOOK:
-                output_list.append(self.output_notebook_path)
```
Comment (kumare3) [inline plugins/flytekit-papermill/flytekitplugins/papermill/task.py:281]: Are you just removing these?

Comment (pingsutw) [reply plugins/flytekit-papermill/flytekitplugins/papermill/task.py:281]: cc @rahul-theorem I know you're using papermill plugin, do you use these paths in the output?

File: plugins/flytekit-papermill/flytekitplugins/papermill/task.py [hunk 4]
```diff
@@ -165,13 +166,14 @@ def __init__(
         if not os.path.exists(self._notebook_path):
             raise ValueError(f"Illegal notebook path passed in {self._notebook_path}")
 
-        if outputs:
+        if outputs and output_notebooks:
```
Comment (wild-endeavor) [inline plugins/flytekit-papermill/flytekitplugins/papermill/task.py:169]:
```suggestion
        if output_notebooks:
```

Comment (wild-endeavor) [reply plugins/flytekit-papermill/flytekitplugins/papermill/task.py:169]: what if you just want a notebook?  don't want to have to make a fake output.

File: plugins/flytekit-papermill/flytekitplugins/papermill/task.py [hunk 5]
```diff
@@ -287,6 +289,8 @@ def execute(self, **kwargs) -> Any:
             else:
                 raise TypeError(f"Expected output {k} of type {type_v} not found in the notebook outputs")
 
+        if len(output_list) == 1:
```
Comment (wild-endeavor) [inline plugins/flytekit-papermill/flytekitplugins/papermill/task.py:292]: what happens if we don't do this?

File: flytekit/core/map_task.py [hunk 6]
```diff
@@ -76,7 +76,11 @@ def __init__(
 
         collection_interface = transform_interface_to_list_interface(actual_task.python_interface, self._bound_inputs)
         self._run_task: PythonFunctionTask = actual_task
-        _, mod, f, _ = tracker.extract_task_module(actual_task.task_function)
+        if issubclass(type(actual_task), PythonInstanceTask):
```
Comment (wild-endeavor) [inline flytekit/core/map_task.py:82]: is there a difference between this and isinstance(actual_task, PythonInstanceTask)

File: flytekit/core/map_task.py [hunk 7]
```diff
@@ -76,7 +76,11 @@ def __init__(
...
+            mod = actual_task.task_type
```
Comment (wild-endeavor) [inline flytekit/core/map_task.py:83]: should this be `actual_task.instantiated_in`?  what happens if there are two of the same notebook tasks, named the same, with the same interface, but in two different .py files?  will there be confusion?

**Review by wild-endeavor — COMMENTED** [review]
the only thing is that this will change the interface for the user right?  we will need to add a disclaimer to the release notes.

File: plugins/flytekit-papermill/flytekitplugins/papermill/task.py (hunk 5 above)
Comment (pingsutw) [reply plugins/flytekit-papermill/flytekitplugins/papermill/task.py:292]: There will be a mismatch between the output type and the downstream task's input type.

File: flytekit/core/map_task.py (hunk 6 above)
Comment (pingsutw) [reply flytekit/core/map_task.py:82]: updated it to `isinstance(actual_task, PythonInstanceTask)`

File: flytekit/core/map_task.py (hunk 7 above)
Comment (pingsutw) [reply flytekit/core/map_task.py:83]: used lhs instead

File: plugins/flytekit-papermill/flytekitplugins/papermill/task.py (hunk 4 above)
Comment (pingsutw) [reply plugins/flytekit-papermill/flytekitplugins/papermill/task.py]: updated it

File: plugins/flytekit-papermill/flytekitplugins/papermill/task.py [hunk 8]
```diff
@@ -165,13 +166,16 @@ def __init__(
         if not os.path.exists(self._notebook_path):
             raise ValueError(f"Illegal notebook path passed in {self._notebook_path}")
 
-        if outputs:
+        if output_notebooks:
+            if outputs is None:
```
Comment (wild-endeavor) [inline plugins/flytekit-papermill/flytekitplugins/papermill/task.py:170]: can do `output = outputs or {}`

**Review by wild-endeavor — APPROVED** [review]

**Pull request merged**